osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,879 @@
+      emit_record = true;
+
+    } else {
+      auto last_file_state = last_file_state_it->second;
+      emit_record = last_file_state != new_record.type;
+    }
+
+    if (emit_record) {
+      if (FLAGS_usn_journal_reader_debug) {
+        TLOG << new_record;
+      }
+
+      new_records.push_back(std::move(new_record));
+      per_file_last_record_type_map[new_record.node_ref_number] =
+          new_record.type;
+
+      if (per_file_last_record_type_map.size() >= 20000U) {
+        auto range_start = per_file_last_record_type_map.begin();
+        auto range_end = std::next(range_start, 10000U);
+
+        per_file_last_record_type_map.erase(range_start, range_end);
+      }
+    }
+  }
+
+  return Status(0);
+}
+
+// TODO: Needs tests; see https://github.com/osquery/osquery/issues/5847
+Status USNJournalReader::ProcessAndAppendUSNRecord(
+    std::vector<USNJournalEventRecord>& record_list,
+    const USN_RECORD* record,
+    USNPerFileLastRecordType& per_file_last_record_type_map,
+    char drive_letter) {
+  // We don't care about range tracking records
+  if (record->MajorVersion == 4U) {
+    return Status(0);
+  }
+
+  if (record->MinorVersion != 0U) {
+    LOG(WARNING) << "Unexpected minor version value";
+  }
+
+  USNJournalEventRecord base_event_record = {};
+  base_event_record.drive_letter = drive_letter;
+
+  base_event_record.journal_record_version =
+      static_cast<size_t>(record->MajorVersion);
+
+  if (!USNParsers::GetUpdateSequenceNumber(
+          base_event_record.update_sequence_number, record)) {
+    return Status(1,
+                  "Failed to get the update sequence number from the record");
+  }
+
+  if (!USNParsers::GetFileReferenceNumber(base_event_record.node_ref_number,
+                                          record)) {
+    return Status(1, "Failed to get the file reference number");
+  }
+
+  if (!USNParsers::GetParentFileReferenceNumber(
+          base_event_record.parent_ref_number, record)) {
+    return Status(1, "Failed to get the parent reference number");
+  }
+
+  if (!USNParsers::GetTimeStamp(base_event_record.record_timestamp, record)) {
+    return Status(1, "Failed to get the timestamp");
+  }
+
+  if (!USNParsers::GetAttributes(base_event_record.attributes, record)) {
+    return Status(1, "Failed to get the file attributes");
+  }
+
+  if (!USNParsers::GetEventString(base_event_record.name, record)) {
+    return Status(1, "Failed to acquire the file name");
+  }
+
+  // Now decompress the record by splitting the `reason` field
+  DWORD reason;
+  if (!USNParsers::GetReason(reason, record)) {
+    return Status(1, "Failed to get the `reason` field from the record");
+  }
+
+  auto status = DecompressRecord(
+      record_list, base_event_record, reason, per_file_last_record_type_map);
+  if (!status.ok()) {
+    return status;
+  }
+
+  return Status(0);
+}
+
+void GetNativeFileIdFromUSNReference(FILE_ID_DESCRIPTOR& file_id,
```

Comment:
I don't understand the use of boostmp , as opposed to using a vector of bytes ... which is what we end up using here anyways.

Reply:
Yeah, I agree. I believe this was an optimization/memory use choice during an early iteration, but the vector is present regardless. I'll refactor.